stdlib/src/handlers/aead_decrypt.rs

```diff
@@ -0,0 +1,178 @@
+        let value =
+            process
+                .get_mem_value(ctx, addr)
+                .ok_or_else(|| AeadDecryptError::MemoryReadFailed {
+                    addr,
+                    reason: alloc::string::String::from("memory read returned None"),
+                })?;
+        ciphertext.push(value);
+    }
+
+    // Read the authentication tag (4 elements at src_ptr + (num_blocks + 1) * 8)
+    let tag_addr = (src_ptr + (num_blocks + 1) * 8) as u32;
+    let mut tag_elements = [miden_core::Felt::ZERO; 4];
+    for i in 0..4 {
+        let addr = tag_addr + i as u32;
+        let value =
+            process
+                .get_mem_value(ctx, addr)
+                .ok_or_else(|| AeadDecryptError::MemoryReadFailed {
+                    addr,
+                    reason: alloc::string::String::from("tag read returned None"),
+                })?;
+        tag_elements[i as usize] = value;
+    }
+
+    // Convert to reference implementation types
+    let secret_key = SecretKey::from_elements(key_word.into());
+    let nonce = Nonce::from(nonce_word);
+    let auth_tag = AuthTag::new(tag_elements);
+
+    // Construct EncryptedData
+    let encrypted_data =
+        EncryptedData::from_parts(DataType::Elements, ciphertext, auth_tag, nonce.clone());
+
+    // Decrypt using the standard reference implementation
+    // This performs tag verification internally
+    let plaintext_with_padding = secret_key.decrypt_elements(&encrypted_data).map_err(|e| {
+        AeadDecryptError::DecryptionFailed {
+            reason: alloc::format!("decryption failed: {:?}", e),
+        }
+    })?;
+
+    // Extract only the data blocks (without padding) to insert into advice
+    // The MASM encrypt procedure will add padding automatically during re-encryption
+    let data_blocks_count = (num_blocks * 8) as usize;
+    let plaintext_data: Vec<_> =
+        plaintext_with_padding.into_iter().take(data_blocks_count).collect();
+
+    // Insert plaintext data (WITHOUT padding) into advice map with key = nonce (4 elements)
+    // The padding will be added by the MASM encrypt procedure during re-encryption
+    // Create an AdviceMap from the iterator
+    let advice_map: miden_core::AdviceMap =
+        core::iter::once((nonce_word, plaintext_data)).collect();
+    let advice_map_mutation = AdviceMutation::extend_map(advice_map);
+
+    Ok(vec![advice_map_mutation])
+}
+
+// ERROR HANDLING
+// ================================================================================================
+
+/// Error types that can occur during AEAD decryption.
+#[derive(Debug, thiserror::Error)]
+enum AeadDecryptError {
+    /// Memory read failed.
+    #[error("failed to read memory at address {addr}: {reason}")]
+    MemoryReadFailed { addr: u32, reason: alloc::string::String },
+
+    /// Decryption failed.
+    #[error("decryption failed: {reason}")]
+    DecryptionFailed { reason: alloc::string::String },
```

Comment:
This is only used once it seems, can we just include the error with `#[source]`?